File: templates/javascript/v1/factorymethod/notifications.js
const NOT_IMPLEMENTED_ERROR = "Not implemented";
const EMAIL_PREFIX = "Email: ";
const SMS_PREFIX = "SMS: ";

export class Notification {
  send(_message) {
    throw new Error(NOT_IMPLEMENTED_ERROR);
  }
}

export class EmailNotification extends Notification {
  send(message) {
    return `${EMAIL_PREFIX}${message}`;
  }
}

export class SmsNotification extends Notification {
  send(message) {
    return `${SMS_PREFIX}${message}`;
  }
}

export class NotificationCreator {
  notify(message) {
    const notification = this.createNotification();
    return notification.send(message);
  }

  createNotification() {
    throw new Error(NOT_IMPLEMENTED_ERROR);
  }
}

export class EmailNotificationCreator extends NotificationCreator {
  createNotification() {
    return new EmailNotification();
  }
}

export class SmsNotificationCreator extends NotificationCreator {
  createNotification() {
    return new SmsNotification();
  }
}

File: templates/python/v1/factorymethod/notifications.py
from __future__ import annotations

from dataclasses import dataclass

EMAIL_PREFIX = "Email: "
SMS_PREFIX = "SMS: "


@dataclass(slots=True)
class EmailNotification:
    def send(self, message: str) -> str:
        return f"{EMAIL_PREFIX}{message}"


@dataclass(slots=True)
class SmsNotification:
    def send(self, message: str) -> str:
        return f"{SMS_PREFIX}{message}"


class NotificationCreator:
    def notify(self, message: str) -> str:
        notification = self.create_notification()
        return notification.send(message)

    def create_notification(self) -> "EmailNotification | SmsNotification":
        raise NotImplementedError


class EmailNotificationCreator(NotificationCreator):
    def create_notification(self):
        return EmailNotification()


class SmsNotificationCreator(NotificationCreator):
    def create_notification(self):
        return SmsNotification()
